### DISCUSSION TITLE
Disable init in interface

### DISCUSSION THREAD

**Discussion by liamlim** [dotnet/csharplang#4030] [General] [answered]
I would like to open a discussion about disabling `init` in interface and abstract classes. What is the point of it? We can only initialize with specific implementation so `init` in interface makes no sense. There are also other consequences. For example consider this

```
interface IA { public int Prop { get; init; } }

class A : IA 
{
  public void SetProp()
  {
    // this is not possible and there is no way to make it possible because there is no 'private init'
    Prop = 5;
  }

  public int Prop { get; init; }
}
```

I don't think that there should be a way for interface to disable private setters in implementing type. I would disable `init` for any abstract property not just for interfaces and abstract classes because implementing type should decide how to construct an object (which constructors to use, whether to use initializers or not).

**Comment by YairHalberstadt** [comment]
One way this is currently useful is when the interface is used as a type constraint:

```csharp
public class C {
    public void M<T>() where T : I, new() {
        var x = new T{ P = 42 };
    }
}

public interface I
{
    public int P { get; init; }
}
```

**Reply by liamlim** [reply]
Honestly I think that this should be solved by enabling something like this:
```
public class C {
    public void M<T>() where T : I, new(int p) {
        var x = new T(42);
    }
}

public interface I
{
    public int P { get; }
}
```

And then all implementators of interface can decide which way they want to create an object. They can only use method `M` if they provide constructor with one int parameter which is fine. And they can also decide whether to provide protected setter or public setter for property P which is NOT possible in current implementation

**Reply by svick** [reply]
That has already been suggested at https://github.com/dotnet/csharplang/discussions/769.

**Comment by svick** [comment, marked as answer]
You *can* implement that interface while having `SetProp`, by using an explicit backing field:

```c#
class A : IA
{
    private int prop;
    
    public void SetProp(int value) => prop = value;
    
    public int Prop { get => prop; init => prop = value; }
}
```

Though I agree it would be nice if you didn't have to do that.

**Reply by YairHalberstadt** [reply]
I've opened https://github.com/dotnet/csharplang/discussions/4031 to discuss that issue.

**Comment by liamlim** [comment]
Imagine a code
```
public abstract class A
{
    public abstract int Prop { get; protected set; }
}
```

Is it possible to enable object initializer for implementing type?
